File: gfg/2024_8_august/c++/18_split_an_array_into_two_equal_sum_subarrays.cpp
// TIME COMPEXITY
// O(n)

// SPACE COMPEXITY
// O(1)

// CODE

class Solution {
  public:
    bool canSplit(vector<int>& arr) {
        int totalSum = 0;
        for(int i = 0; i < arr.size(); i++){
            totalSum += arr[i];
        }
        
        int sum = 0;
        for(int i= 0; i < arr.size(); i++){
            sum += arr[i];
            if(sum * 2 == totalSum) return true;
        }
        return false;
    }
};

File: gfg/2024_8_august/python/18_split_an_array_into_two_equal_sum_subarrays.py
# CODE

class Solution:
    def canSplit(self, arr):
        total_sum = sum(arr)
        if total_sum % 2 != 0:
            return False
        prefix_sum = 0
        suffix_sum = total_sum
        for i in range(len(arr) - 1):
            prefix_sum += arr[i]
            suffix_sum -= arr[i]
            if prefix_sum == suffix_sum:
                return True
        return False